pkg/reconciler/cache/replication/replication_reconcile_unstructured.go

```diff
@@ -0,0 +1,213 @@
+		// TODO: in the future the original RV will have to be stored in an annotation (?)
+		// so that the clients that need to modify the original/local object can do it
+		unstructuredLocalObject.SetResourceVersion("")
+		_, err := c.dynamicCacheClient.Cluster(cluster).Resource(*gvr).Namespace(unstructuredLocalObject.GetNamespace()).Create(ctx, unstructuredLocalObject, metav1.CreateOptions{})
+		return err
+	}
+
+	metaChanged, err := ensureUnstructuredMeta(unstructuredCacheObject, unstructuredLocalObject)
+	if err != nil {
+		return err
+	}
+	remainingChanged, err := ensureUnstructuredRemaining(unstructuredCacheObject, unstructuredLocalObject)
+	if err != nil {
+		return err
+	}
+	if !metaChanged && !remainingChanged {
+		return nil
+	}
+
+	if metaChanged || remainingChanged {
+		_, err := c.dynamicCacheClient.Cluster(cluster).Resource(*gvr).Namespace(unstructuredCacheObject.GetNamespace()).Update(ctx, unstructuredCacheObject, metav1.UpdateOptions{})
+		return err
+	}
+	return nil
+}
+
+func (c *controller) handleUnstructuredObjectDeletion(ctx context.Context, cluster logicalcluster.Name, gvr *schema.GroupVersionResource, unstructuredCacheObject *unstructured.Unstructured) error {
+	if unstructuredCacheObject == nil {
+		return nil // the cached object already removed
+	}
+	if unstructuredCacheObject.GetDeletionTimestamp() == nil {
+		return c.dynamicCacheClient.Cluster(cluster).Resource(*gvr).Delete(ctx, unstructuredCacheObject.GetName(), metav1.DeleteOptions{})
+	}
+	return nil
+}
+
+// ensureUnstructuredMeta changes unstructuredCacheObject's metadata to match unstructuredLocalObject's metadata except the ResourceVersion and the shard annotation fields
+func ensureUnstructuredMeta(unstructuredCacheObject *unstructured.Unstructured, unstructuredLocalObject *unstructured.Unstructured) (metaChanged bool, err error) {
+	unstructuredCacheObjectMetaRaw, hasUnstructuredCacheObjectMetaRaw, err := unstructured.NestedFieldNoCopy(unstructuredCacheObject.Object, "metadata")
+	if err != nil {
+		return false, err
+	}
+	unstructuredCacheObjectMeta, ok := unstructuredCacheObjectMetaRaw.(map[string]interface{})
+	if !ok {
+		return false, fmt.Errorf("metadata field of unstructuredCacheObject is of the type %T, expected map[string]interface{}", unstructuredCacheObjectMetaRaw)
+	}
+	unstructuredLocalObjectMetaRaw, hasUnstructuredLocalObjectMetaRaw, err := unstructured.NestedFieldNoCopy(unstructuredLocalObject.Object, "metadata")
+	if err != nil {
+		return false, err
+	}
+	unstructuredLocalObjectMeta, ok := unstructuredLocalObjectMetaRaw.(map[string]interface{})
+	if !ok {
+		return false, fmt.Errorf("metadata field of unstructuredLocalObjectMeta is of the type %T, expected map[string]interface{}", unstructuredLocalObjectMetaRaw)
+	}
+	if !hasUnstructuredLocalObjectMetaRaw && !hasUnstructuredCacheObjectMetaRaw {
+		return false, nil // no-op
+	}
+	if !hasUnstructuredLocalObjectMetaRaw {
+		unstructured.RemoveNestedField(unstructuredCacheObject.Object, "metadata")
```

Comment:
is unstructuredCacheObject from a lister? We cannot touch it, not even temporarily.